plugins/out_s3/s3.c

```diff
@@ -1755,198 +1866,1579 @@ static void s3_upload_queue(struct flb_config *config, void *out_context)
+
+    valid_path = (char *) path;
+
+    while (*valid_path == '.' ||
+           *valid_path == '/') {
+            valid_path++;
+    }
+
+    url = flb_sds_create(context->authorization_endpoint_url);
+
+    if (url == NULL) {
+        return -1;
+    }
+
+    tmp = flb_sds_printf(&url, "/put_object_presigned_url/%s/%s/%s", bucket, tag, valid_path);
+
+    if (tmp != NULL) {
+        url = tmp;
+
+        ret = blob_fetch_pre_signed_url(context, result_url, (char *) url);
+    }
+    else {
+        ret = -1;
+    }
+
+    flb_sds_destroy(url);
+
+    return ret;
+}
+
+static int blob_fetch_create_multipart_upload_pre_signed_url(struct flb_s3 *context,
+                                                             flb_sds_t *result_url,
+                                                             char *tag,
+                                                             char *bucket,
+                                                             char *path)
+{
+    char      *valid_path;
+    int        ret;
+    flb_sds_t  url;
+    flb_sds_t  tmp;
+
+    valid_path = (char *) path;
+
+    while (*valid_path == '.' ||
+           *valid_path == '/') {
+            valid_path++;
+    }
+
+    url = flb_sds_create(context->authorization_endpoint_url);
+
+    if (url == NULL) {
+        return -1;
+    }
+
+    tmp = flb_sds_printf(&url, "/multipart_creation_presigned_url/%s/%s/%s", bucket, tag, valid_path);
+
+    if (tmp != NULL) {
+        url = tmp;
+
+        ret = blob_fetch_pre_signed_url(context, result_url, (char *) url);
+    }
+    else {
+        ret = -1;
+    }
+
+    flb_sds_destroy(url);
+
+    return ret;
+}
+
+static int blob_fetch_multipart_upload_pre_signed_url(struct flb_s3 *context,
+                                                             flb_sds_t *result_url,
+                                                             char *tag,
+                                                             char *bucket,
+                                                             char *path,
+                                                             char *upload_id,
+                                                             int part_number)
+{
+    char      *valid_path;
+    int        ret;
+    flb_sds_t  url;
+    flb_sds_t  tmp;
+
+    valid_path = (char *) path;
+
+    while (*valid_path == '.' ||
+           *valid_path == '/') {
+            valid_path++;
+    }
+
+    url = flb_sds_create(context->authorization_endpoint_url);
+
+    if (url == NULL) {
+        return -1;
+    }
+
+    tmp = flb_sds_printf(&url, "/multipart_upload_presigned_url/%s/%s/%s/%s/%d", bucket, tag, valid_path, upload_id, part_number);
+
+    if (tmp != NULL) {
+        url = tmp;
+
+        ret = blob_fetch_pre_signed_url(context, result_url, (char *) url);
+    }
+    else {
+        ret = -1;
+    }
+
+    flb_sds_destroy(url);
+
+    return ret;
+}
+
+static int blob_fetch_multipart_complete_pre_signed_url(struct flb_s3 *context,
```

Comment:
These fetch_*_pre_signed_url methods are near duplicates with minor changes to the tmp (path variable), can we consider using a helper method for getting the presigned url and have each method handle their own tmp/path generation to reduce code?

Reply:
Makes sense, I'll address it.